utils/src/main/scala/za/co/absa/enceladus/utils/validation/SchemaPathValidator.scala

```diff
@@ -141,54 +141,61 @@ object SchemaPathValidator {
-          return Seq(ValidationError(s"Case insensitive variant of a cloumn '$parentPath$currentField' already exists so it cannot be used as an output column '$fullPath'."))
+    else {
+
+      val currentField = path(0)
+      val fullPath = s"$parentPath${path.mkString(".")}"
+      if (parentOnly && path.length == 1) {
+        val matched = checkMatchType(schema, currentField)
+        matched match {
+          case ExactMatch(_) if fullPathNew =>
+            Seq(ValidationError(s"Column '$parentPath$currentField' already exists so it cannot be used as an output column '$fullPath'."))
+          case CaseInsensitiveMatch(_) if fullPathNew =>
+            Seq(ValidationError(s"Case insensitive variant of a cloumn '$parentPath$currentField' already exists so it cannot be used as an output column '$fullPath'."))
+          case _ => Nil
         }
-      }
-      return Nil
-    }
+      } else {
 
-    val exactMatch = schema.fields.find(field => field.name == currentField)
-    val failures = exactMatch match {
-      case Some(field) =>
-        val dataType = getUnderlyingType(field.dataType)
-        dataType match {
-          case st: StructType =>
-            validateSchemaPathArray(st, path.drop(1), parentOnly, fullPathNew, s"$parentPath${field.name}.")
-          case _ =>
-            if (path.length > 1) {
-              Seq(ValidationError(s"Column '$parentPath$currentField' is a primitive type and can't contain child fields '$fullPath'."))
-            } else {
-              Nil
+        val matched = checkMatchType(schema, currentField)
+        val failures = matched match {
+          case ExactMatch(field) =>
+            val dataType = getUnderlyingType(field.dataType)
+            dataType match {
+              case st: StructType =>
+                validateSchemaPathArray(st, path.drop(1), parentOnly, fullPathNew, s"$parentPath${field.name}.")
+              case _ if path.length > 1 =>
+                Seq(ValidationError(s"Column '$parentPath$currentField' is a primitive type and can't contain child fields '$fullPath'."))
+              case _ => Nil
             }
-        }
-      case None =>
-        val caseInsensitiveMatch = schema.fields.find(field => field.name.compareToIgnoreCase(currentField) == 0)
-        caseInsensitiveMatch match {
-          case Some(field) =>
+          case CaseInsensitiveMatch(field) =>
             Seq(ValidationError(s"Column name '$parentPath$currentField' does not case-sensitively match '$parentPath${field.name}'."))
-          case None =>
+          case _ =>
             Seq(ValidationError(s"Column name '$parentPath$currentField' does not exist."))
         }
+        failures
+      }
+    }
+  }
+
+  trait Match
+
+  case class ExactMatch(field: StructField) extends Match
+
+  case class CaseInsensitiveMatch(field: StructField) extends Match
+
+  case object NoMatch extends Match
+
+  private def checkMatchType(schema: StructType, currentField: String): Match = {
```

Comment:
I am wondering if the code got simpler if this method's return type was actually `Option[Match]` and there would be no `NoMatch` case object, just the two actual match types.

I have a hunch that the processing of such a result could then be done in 
```
matched.map{actualMatch => ... }.toSeq
```

Reply:
I actually thought of doing the same the first time, but I thought that nesting of the return type when it's possible not to would be better

Reply:
![](https://user-images.githubusercontent.com/4457378/78889662-67804800-7a64-11ea-8753-1da8fb355a39.jpg)
(Futurama reference)